controllers/userController.js

```diff
@@ -277,52 +276,130 @@ exports.getAccountProfile = function(req, res) {
  * POST /account/profile
  * Update profile information.
  */
-
  exports.postUpdateProfile = function(req, res, next) {
+
+	 // to add a key
+	 // 1.sanitize
+	 // 2. check for key in req.body
+	 // 	3. req.assert validation by key
+	 // 	4. add req.body.key to updateFields.key
+	 // 5. add key to copyKeysIfExist
+	 
+	var mustConfirmPassword = false
+
 	req.sanitize('name').trim()
 	req.sanitize('email').trim()
- 	req.assert('email', 'Email is not valid').isEmail()
- 	req.assert('name', 'Name is not valid').notEmpty()
+	req.sanitize('profile[website]').trim()
+	req.sanitize('profile[bio]').trim()
+
+	var updateFields = {
+		profile: {},
+		preferences: {}
+	}
+
+	if ('email' in req.body) {
+		mustConfirmPassword = true
+		req.assert('email', 'Email is not valid').isEmail()
+		updateFields.email = req.body.email
+	}
+
+	if('password' in req.body) {
+		mustConfirmPassword = true
+		req.assert('password', 'New password is not valid').isLength({min: 6})
+		updateFields.password = req.body.password
+	}
+
+	if ('name' in req.body) {
+		req.assert('name', 'Name is not valid').notEmpty()
+		updateFields.name = req.body.name.replace(/[<>\\\\'\"]+/gim, '')
+	}
+
+	if (req.body.profile) {
+		if ('website' in req.body.profile) {
+			// allow '' or valid URL only
+			if (req.body.profile.website !== '')
+				req.assert('profile[website]', 'Website is not valid').isURL({require_protocol:true})
+			updateFields.profile.website = req.body.profile.website
+		}
+
+		if ('bio' in req.body.profile) {
+			req.assert('profile[bio]', 'Bio must be shorter than 200 characters').isLength({min: 0, max: 200})
+			updateFields.profile.bio = req.body.profile.bio
+		}
+	}
+
+	 if (req.body.preferences) {
+		 if ('publishDefaultPublic' in req.body.preferences)
+			 updateFields.preferences.publishDefaultPublic = (req.body.preferences.publishDefaultPublic === '1')
+	 }
+
+	if (mustConfirmPassword)
+		req.assert('current_password', 'Your current password is required').notEmpty()
 
  	var errors = req.validationErrors()
- 	var wantJson = req.xhr || req.path.slice(-5) === '.json'
 
  	if (errors) {
- 		if (wantJson) {
- 			return res.status(400).json(
-				helper.responseStatusError('Failed validation', errors, {request:req.body})
-			)
- 		} else {
- 			req.flash('errors', helper.parseErrors(errors))
- 			return res.redirect('/account')
- 		}
+		delete req.body.password
+		delete req.body.current_password
+		return helper.respond(req, res, 400, 'Failed validation', errors, null, '/account/profile')
  	}
 
- 	User.findById(req.user.id, function(err, user) {
- 		if (err) return next(err)
+	 function updateUser(user, updateFields) {
+		function copyKeysIfExist(keys, fromObj, toObj) {
+			keys.forEach(function(key){
+				if (key in fromObj)
+					toObj[key] = fromObj[key]
+			})
+		}
+		copyKeysIfExist(['email', 'password', 'name'], updateFields , user)
+		copyKeysIfExist(['website', 'bio'] , updateFields.profile, user.profile)
```

Comment:
FYI, mongoose will drop any extra keys from the object based on the model rules, so you could just `_.assign()`.. or do this.